distributed/stealing.py

```diff
@@ -344,7 +339,10 @@ def maybe_move_task(level, ts, sat, idl, duration, cost_multiplier):
                             thieves = idle
                         if not thieves:
                             break
-                        thief = thieves[i % len(thieves)]
+
+                        thief = min(
+                            thieves, key=partial(self.scheduler.worker_objective, ts)
+                        )
```

Comment:
How many thieves can there be here?  If the answer is "maybe a lot" then we might need to be careful here.

Reply:
I'm also curious how much impact this choice in particular had on the workload we're testing against.

Reply:
> How many thieves can there be here? If the answer is "maybe a lot" then we might need to be careful here.

Worst case situation would be if all but one worker are idling. then we'd have `W-1` potential thieves. I thought about sampling this down to 20 or smth like that to avoid any extreme cases. Sampling and picking the best of the sample still sounds better than random. However...

> I'm also curious how much impact this choice in particular had on the workload we're testing against.

Nothing, at all. This is the one chance which didn't actually impact this particular example. It just _felt_ right? idk. That's one of the things I want to test on different workloads

Reply:
In particular for workloads with actual payload data. If the dependents don't have data, this is a rather extreme stealing edge case and I'd like to verify that this also behaves nicely for proper payloads

Reply:
In decide_worker we choose the best worker if there are less than 20 workers and we choose an incrementing worker if there are more than 20 workers.  

In practice though I doubt that it matters much.  The difference in occupancy between thieves and victims means that we're making the world significantly better with the dumb change.  If this results in a suboptimal world that's ok... we'll just improve it again in a future cycle if it becomes a problem.

Reply:
> In decide_worker we choose the best worker if there are less than 20 workers and we choose an incrementing worker if there are more than 20 workers.

I also thought about just using decide_worker. After all, why would the logic be any different (given that decide_worker takes a whitelist <-> thieves). However, the important logic is in the objective function which is why I picked that. If this things turns out to be valuable, we can have a closer look

> In practice though I doubt that it matters much. The difference in occupancy between thieves and victims means that we're making the world significantly better with the dumb change. 

True. As I said, the current problem doesn't actually trigger anything where this would matter. If I can not prove an improvement, I'll remove it again.

> The difference in occupancy

The only thing I would like to leverage here is to pick a thief with dependencies, if that exists. that's implicitly done by the worker objective but could also be made explicit. My gut tells me this can make a difference but you might be right that this is only marginal.

> we'll just improve it again in a future cycle if it becomes a problem.

There will not be another cylce to correct this unless the worker becomes saturated soon-ish, will there?. Non-saturated workers are not victimized.

Reply:
> There will not be another cylce to correct this unless the worker becomes saturated soon-ish, will there?. Non-saturated workers are not victimized.

Correct, but the stolen task probably won't be run immediately anyway.  It's probably at the end of the queue.  As the computation continues progressing and some workers become more free before others the saturated/non-saturated split will grow and trigger another cycle of transfers if necessary.